### ISSUE TITLE
EvalDetection may not find best pairings of ground truth / predicted

### ISSUE THREAD

**Issue by k-allen** [biometrics/openbr#212]
In the function's while loop (pasted below), we’re pulling out a ground truth value and then removing its closest match from the predicted list.  It seems possible to me that this will remove a prediction that’s closer in value to another ground truth item (and thus probably corresponded to that item).  

Say detections.truth has A and B, and detections.predicted has P1 and P2.  If P1 is right on top of B, but P2 is a false positive, P1 may still be closer to A than P2 is.  So we’ll end up removing it from consideration before we start looking for a match to B—pulling out a potentially good prediction because a ground truth value early in the truth gallery was a false negative.

Anyone know if this was already considered when the function was created?  If it's an issue, one way we could address it would be to only (or first) consider overlaps that are mutual -- i.e. each is the best overlap for the other.  I believe that would avoid the problem noted above and would have the side effect of ignoring some stray false positives and false negatives as potentially overlapping with half of a matched pair.  

``` C++
float EvalDetection(const QString &predictedGallery, const QString &truthGallery, const QString &csv)
{
    qDebug("Evaluating detection of %s against %s", qPrintable(predictedGallery), qPrintable(truthGallery));

    // Organized by file, QMap used to preserve order
    QMap<QString, Detections> allDetections = getDetections(predictedGallery, truthGallery);

    QList<ResolvedDetection> resolvedDetections, falseNegativeDetections;
    int totalTrueDetections = 0;
    foreach (Detections detections, allDetections.values()) { // For every file
        totalTrueDetections += detections.truth.size();

        // Try to associate ground truth detections with predicted detections
        while (!detections.truth.isEmpty() && !detections.predicted.isEmpty()) {
            const Detection truth = detections.truth.takeFirst(); // Take removes the detection
            int bestIndex = -1;
            float bestOverlap = -std::numeric_limits<float>::max();
            // Find the nearest predicted detection to this ground truth detection
            for (int i=0; i<detections.predicted.size(); i++) {
                const float overlap = truth.overlap(detections.predicted[i]);
                if (overlap > bestOverlap) {
                    bestOverlap = overlap;
                    bestIndex = i;
                }
            }
            // Removing the detection prevents us from considering it twice.
            // We don't want to associate two ground truth detections with the
            // same prediction, over vice versa.
            const Detection predicted = detections.predicted.takeAt(bestIndex);
            resolvedDetections.append(ResolvedDetection(predicted.confidence, bestOverlap));
        }

        foreach (const Detection &detection, detections.predicted)
            resolvedDetections.append(ResolvedDetection(detection.confidence, 0));
        for (int i=0; i<detections.truth.size(); i++)
            falseNegativeDetections.append(ResolvedDetection(-std::numeric_limits<float>::max(), 0));
    }

    std::sort(resolvedDetections.begin(), resolvedDetections.end());

    QStringList lines;
    lines.append("Plot, X, Y");
    lines.append(computeDetectionResults(resolvedDetections, totalTrueDetections, true));
    lines.append(computeDetectionResults(resolvedDetections, totalTrueDetections, false));

    float averageOverlap;
    { // Overlap Density
        QList<ResolvedDetection> allDetections; allDetections << resolvedDetections << falseNegativeDetections;
        const int keep = qMin(allDetections.size(), Max_Points);
        lines.reserve(lines.size() + keep);
        float totalOverlap = 0;
        for (int i=0; i<keep; i++) {
            const float overlap = allDetections[double(i) / double(keep-1) * double(allDetections.size()-1)].overlap;
            totalOverlap += overlap;
            lines.append(QString("Overlap,%1,1").arg(QString::number(allDetections[double(i) / double(keep-1) * double(allDetections.size()-1)].overlap)));
        }
        averageOverlap = totalOverlap / keep;
    }
    lines.append(QString("AverageOverlap,%1,").arg(QString::number(averageOverlap)));

    QtUtils::writeFile(csv, lines);
    qDebug("Average Overlap = %.3f", averageOverlap);
    return averageOverlap;
}
```
